Support generics

For example,

```rust
#[derive(Debug, IntoEnum)]
pub enum Value<A: Scalar> {
    Scalar(A),
    Vector(Array<A, Ix1>),
}
```

this code does not work:

```output
error[E0412]: cannot find type `A` in this scope
  --> main.rs:17:17
   |
17 | #[derive(Debug, IntoEnum)]
   |                 ^^^^^^^^ not found in this scope
```

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -112,7 +112,6 @@ fn impl_newtype_deref(ast: &syn::MacroInput) -> quote::Tokens {
             fn deref_mut(&mut self) -> &mut Self::Target { &mut self.0 }
         }
     }
-
 }
 
 fn impl_error(ast: &syn::MacroInput) -> quote::Tokens {

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -9,11 +9,9 @@ use proc_macro::TokenStream;
 #[proc_macro_derive(EnumError)]
 pub fn enum_error(input: TokenStream) -> TokenStream {
     let ast = into_ast(input);
-    let name = &ast.ident;
-    let variants = get_enum_variants(&ast);
-    let froms = impl_into_enum(&name, &variants);
-    let display = impl_display(&name, &variants);
-    let error = impl_error(&name, &variants);
+    let froms = impl_into_enum(&ast);
+    let display = impl_display(&ast);
+    let error = impl_error(&ast);
     let tokens = quote!{ #froms #display #error };
     tokens.parse().unwrap()
 }
@@ -21,18 +19,14 @@ pub fn enum_error(input: TokenStream) -> TokenStream {
 #[proc_macro_derive(IntoEnum)]
 pub fn into_enum(input: TokenStream) -> TokenStream {
     let ast = into_ast(input);
-    let name = &ast.ident;
-    let variants = get_enum_variants(&ast);
-    impl_into_enum(&name, &variants).parse().unwrap()
+    impl_into_enum(&ast).parse().unwrap()
 }
 
 #[proc_macro_derive(NewType)]
 pub fn newtype(input: TokenStream) -> TokenStream {
     let ast = into_ast(input);
-    let name = &ast.ident;
-    let field = get_basetype(&ast);
-    let from = impl_newtype_from(&name, &field);
-    let deref = impl_newtype_deref(&name, &field);
+    let from = impl_newtype_from(&ast);
+    let deref = impl_newtype_deref(&ast);
     let tokens = quote!{ #from #deref };
     tokens.parse().unwrap()
 }
@@ -67,7 +61,10 @@ fn get_basetype(ast: &syn::MacroInput) -> &syn::Field {
     }
 }
 
-fn impl_into_enum(name: &syn::Ident, variants: &Vec<syn::Variant>) -> quote::Tokens {
+fn impl_into_enum(ast: &syn::MacroInput) -> quote::Tokens {
+    let name = &ast.ident;
+    let variants = get_enum_variants(&ast);
+    let (impl_generics, ty_generics, where_clause) = ast.generics.split_for_impl();
     let impls = variants.iter().map(|var| {
         let v = &var.ident;
         let cont = match var.data {
@@ -77,7 +74,7 @@ fn impl_into_enum(name: &syn::Ident, variants: &Vec<syn::Variant>) -> quote::Tok
         assert!(cont.len() == 1, "Single Tuple is required");
         let ctype = &cont[0].ty;
         quote!{
-            impl From<#ctype> for #name {
+            impl #impl_generics From<#ctype> for #name #ty_generics #where_clause {
                 fn from(val: #ctype) -> Self {
                     #name::#v(val)
                 }
@@ -87,10 +84,13 @@ fn impl_into_enum(name: &syn::Ident, variants: &Vec<syn::Variant>) -> quote::Tok
     quote!{ #(#impls)* }
 }
 
-fn impl_newtype_from(name: &syn::Ident, field: &syn::Field) -> quote::Tokens {
+fn impl_newtype_from(ast: &syn::MacroInput) -> quote::Tokens {
+    let name = &ast.ident;
+    let field = get_basetype(&ast);
     let base = &field.ty;
+    let (impl_generics, ty_generics, where_clause) = ast.generics.split_for_impl();
     quote!{
-        impl From<#base> for #name {
+        impl #impl_generics From<#base> for #name #ty_generics #where_clause {
             fn from(val: #base) -> Self {
                 #name(val)
             }
@@ -98,27 +98,33 @@ fn impl_newtype_from(name: &syn::Ident, field: &syn::Field) -> quote::Tokens {
     }
 }
 
-fn impl_newtype_deref(name: &syn::Ident, field: &syn::Field) -> quote::Tokens {
+fn impl_newtype_deref(ast: &syn::MacroInput) -> quote::Tokens {
+    let name = &ast.ident;
+    let field = get_basetype(&ast);
     let base = &field.ty;
+    let (impl_generics, ty_generics, where_clause) = ast.generics.split_for_impl();
     quote!{
-        impl ::std::ops::Deref for #name {
+        impl #impl_generics ::std::ops::Deref for #name #ty_generics #where_clause {
             type Target = #base;
             fn deref(&self) -> &Self::Target { &self.0 }
         }
-        impl ::std::ops::DerefMut for #name {
+        impl #impl_generics ::std::ops::DerefMut for #name #ty_generics #where_clause {
             fn deref_mut(&mut self) -> &mut Self::Target { &mut self.0 }
         }
     }
 
 }
 
-fn impl_error(name: &syn::Ident, variants: &Vec<syn::Variant>) -> quote::Tokens {
+fn impl_error(ast: &syn::MacroInput) -> quote::Tokens {
+    let name = &ast.ident;
+    let variants = get_enum_variants(&ast);
+    let (impl_generics, ty_generics, where_clause) = ast.generics.split_for_impl();
     let snips = variants.iter().map(|var| {
         let v = &var.ident;
         quote!{ #name::#v(ref err) => err.description() }
     });
     quote!{
-        impl ::std::error::Error for #name {
+        impl #impl_generics ::std::error::Error for #name #ty_generics #where_clause {
             fn description(&self) -> &str {
                 match *self {
                     #(#snips), *
@@ -128,13 +134,16 @@ fn impl_error(name: &syn::Ident, variants: &Vec<syn::Variant>) -> quote::Tokens 
     }
 }
 
-fn impl_display(name: &syn::Ident, variants: &Vec<syn::Variant>) -> quote::Tokens {
+fn impl_display(ast: &syn::MacroInput) -> quote::Tokens {
+    let name = &ast.ident;
+    let variants = get_enum_variants(&ast);
+    let (impl_generics, ty_generics, where_clause) = ast.generics.split_for_impl();
     let snips = variants.iter().map(|var| {
         let v = &var.ident;
         quote!{ #name::#v(ref err) => err.fmt(f) }
     });
     quote!{
-        impl ::std::fmt::Display for #name {
+        impl #impl_generics ::std::fmt::Display for #name #ty_generics #where_clause {
             fn fmt(&self, f: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {
                 match *self {
                     #(#snips), *

--- a/tests/newtype.rs
+++ b/tests/newtype.rs
@@ -24,9 +24,9 @@ fn newtype() {
     n.func_mut(); // test DerefMut
 }
 
-struct BaseT<A> {
-    a: A,
+pub struct BaseT<A> {
+    pub a: A,
 }
 
 #[derive(NewType)]
-struct NewT<A>(Base<A>);
+pub struct NewT<A>(BaseT<A>);

--- a/tests/newtype.rs
+++ b/tests/newtype.rs
@@ -23,3 +23,10 @@ fn newtype() {
     n.func(); // test Deref
     n.func_mut(); // test DerefMut
 }
+
+struct BaseT<A> {
+    a: A,
+}
+
+#[derive(NewType)]
+struct NewT<A>(Base<A>);

